crates/bevy_sprite/src/frustum_culling.rs

```diff
@@ -0,0 +1,126 @@
+use bevy_transform::components::Transform;
+use bevy_window::Windows;
+
+use crate::{Sprite, TextureAtlas, TextureAtlasSprite};
+
+struct Rect {
+    position: Vec2,
+    size: Vec2,
+}
+
+impl Rect {
+    #[inline]
+    pub fn is_overlapping(&self, other: Rect) -> bool {
+        !(self.position.x > other.end().x
+            || self.end().x < other.position.x
+            || self.position.y > other.end().y
+            || self.end().y < other.position.y)
+    }
+
+    #[inline]
+    pub fn end(&self) -> Vec2 {
+        self.position + self.size
+    }
+}
+pub fn frustum_culling_sprites(
+    commands: &mut Commands,
+    windows: Res<Windows>,
+    cameras: Query<&Transform, With<Camera>>,
+    visible: Query<&Frustum, With<Sprite>>,
+    sprites: Query<(Entity, &Transform, &Sprite)>,
+) {
+    let window = windows.get_primary().unwrap();
+    let window_size = Vec2::new(window.width(), window.height());
+
+    for camera_transform in cameras.iter() {
+        let camera_position = Vec2::new(
+            camera_transform.translation.x,
+            camera_transform.translation.y,
+        );
+
+        let camera_size = window_size * camera_transform.scale.truncate();
+
+        let rect = Rect {
+            position: camera_position - camera_size / Vec2::new(2.0, 2.0),
+            size: camera_size,
+        };
+
+        for (entity, drawable_transform, sprite) in sprites.iter() {
+            let sprite_rect = Rect {
+                position: drawable_transform.translation.truncate()
+                    - sprite.size / Vec2::new(2.0, 2.0),
+                size: sprite.size,
+            };
+
+            if rect.is_overlapping(sprite_rect) {
+                if visible.get(entity).is_err() {
+                    commands.insert_one(entity, Frustum::default());
+                }
+            } else if visible.get(entity).is_ok() {
+                commands.remove_one::<Frustum>(entity);
+            }
+        }
+    }
+}
+
+pub fn frustum_culling_atlases(
+    commands: &mut Commands,
+    windows: Res<Windows>,
+    textures: Res<Assets<TextureAtlas>>,
+    cameras: Query<&Transform, With<Camera>>,
+    visible: Query<&Frustum, With<TextureAtlasSprite>>,
+    sprites: Query<(
+        Entity,
+        &Transform,
+        &TextureAtlasSprite,
+        &Handle<TextureAtlas>,
+    )>,
+) {
+    let window = windows.get_primary().unwrap();
+    let window_size = Vec2::new(window.width(), window.height());
+
+    for camera_transform in cameras.iter() {
+        let camera_position = Vec2::new(
+            camera_transform.translation.x,
+            camera_transform.translation.y,
+        );
+
+        let camera_size = window_size * camera_transform.scale.truncate();
+
+        let rect = Rect {
+            position: camera_position - camera_size / Vec2::new(2.0, 2.0),
+            size: camera_size,
+        };
+
+        for (entity, drawable_transform, sprite, atlas_handle) in sprites.iter() {
```

Comment:
This could also be turned into a parallel iterator.

Reply:
done

Reply:
Given that this a regular iterator, I assume you backed out of this change @Byteron?